// appium-installer/dist/serverInstall.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDriver = getDriver;
exports.installAppiumServer = installAppiumServer;
exports.installDrivers = installDrivers;
exports.installPlugin = installPlugin;
exports.runAppiumDoctor = runAppiumDoctor;
var _inquirer = _interopRequireDefault(require("inquirer"));
var _ora = _interopRequireDefault(require("ora"));
var _plugin = require("./plugin.js");
var _logger = _interopRequireDefault(require("./logger.js"));
var _shelljs = _interopRequireDefault(require("shelljs"));
var _chalk = _interopRequireDefault(require("chalk"));
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const ui = new _logger.default().getInstance();
// @ts-ignore

async function validateAndInstallUserSelectedServerVersion() {
  const {
    enteredVersion
  } = await _inquirer.default.prompt([{
    type: 'input',
    name: 'enteredVersion',
    message: 'Enter your Server version number: '
  }]);
  _shelljs.default.exec(`npm view 'appium@*' versions`, {
    silent: true
  }).includes(enteredVersion) ? await installAppiumServerWithVersion(enteredVersion) : await retryAppiumServerInstall();
}
async function retryAppiumServerInstall() {
  newLine();
  ui.log.write(_chalk.default.red(`Please try again with a valid version number...`));
  await installAppiumServer();
}
async function installAppiumServer() {
  newLine();
  const options = [{
    name: 'Select latest Server version: ',
    fn: installAppiumServerWithVersion,
    value: 1
  }, {
    name: 'Select custom Server version: ',
    fn: validateAndInstallUserSelectedServerVersion,
    value: 2
  }];
  const {
    selectedChoice
  } = await _inquirer.default.prompt([{
    type: 'list',
    name: 'selectedChoice',
    message: 'Choose your version: ',
    choices: options.map(option => option.name)
  }]);
  const currentOption = options.find(option => option.name === selectedChoice);
  currentOption.value === 1 ? await currentOption.fn('latest') : await currentOption.fn();
  ui.log.write('\n');
}
async function installAppiumServerWithVersion(selectedVersion) {
  const spinner = (0, _ora.default)('Installing Appium Server').start();
  try {
    _shelljs.default.exec(`npm install -g appium@${selectedVersion}`);
    const {
      stdout
    } = _shelljs.default.exec('appium -v');
    spinner.succeed(`💥 💥 💥 Successfully installed server version ${stdout}`);
  } catch (err) {
    spinner.fail(err);
    throw new Error(err);
  } finally {
    spinner.stop();
  }
  newLine();
}
async function getDriver() {
  newLine();
  let drivers = [];
  const spinner = (0, _ora.default)('Fetching available official drivers').start();
  try {
    const {
      stdout
    } = _shelljs.default.exec('appium driver list --json');
    Object.keys(JSON.parse(stdout)).forEach(value => drivers.push({
      name: value
    }));
    spinner.succeed();
    return drivers;
  } catch (err) {
    spinner.fail(err);
    spinner.stop();
  } finally {
    spinner.stop();
  }
}
async function installDrivers(driverName) {
  newLine();
  const driverVersion = await _inquirer.default.prompt([{
    type: 'checkbox',
    message: `Install ${driverName} driver version`,
    name: 'drivers',
    choices: ['latest', 'custom']
  }]);
  await appiumDriverInstallChoice(driverVersion, driverName);
}
async function appiumDriverInstallChoice(driverVersion, driverName) {
  if (driverVersion.drivers[0] === 'latest') {
    console.log(`Installing latest ${driverName} driver...`);
    await _shelljs.default.exec(`appium driver uninstall ${driverName}`);
    await _shelljs.default.exec(`appium driver install ${driverName}`);
  }
  if (driverVersion.drivers[0] === 'custom') {
    console.log(`Installing custom ${driverName} driver...`);
    const {
      enteredVersion
    } = await _inquirer.default.prompt([{
      type: 'input',
      name: 'enteredVersion',
      message: `Enter the ${driverName} version number: `
    }]);
    let driverStr = null;
    const driverStrFirst = "npm view 'appium";
    const driverStrSecond = "driver@*' versions";
    switch (driverName) {
      case 'espresso':
        driverStr = driverStrFirst + '-espresso-' + driverStrSecond;
        break;
      case 'uiautomator2':
        driverStr = driverStrFirst + '-uiautomator2-' + driverStrSecond;
        break;
      case 'mac2':
        driverStr = driverStrFirst + '-mac2-' + driverStrSecond;
        break;
      case 'gecko':
        driverStr = driverStrFirst + '-gecko' + driverStrSecond;
        break;
      case 'chromium':
        driverStr = driverStrFirst + '-chromium-' + driverStrSecond;
        break;
      case 'safari':
        driverStr = driverStrFirst + '-safari-' + driverStrSecond;
        break;
      case 'xcuitest':
        driverStr = driverStrFirst + '-xcuitest-' + driverStrSecond;
        break;
    }
    (await _shelljs.default.exec(`${driverStr}`, {
      silent: true
    }).includes(enteredVersion)) ? await installSelectedDriverWithSpecificVersion(driverName, enteredVersion) : await retrySelectedDriverInstall(driverName);
  }
}
async function installSelectedDriverWithSpecificVersion(driverName, enteredVersion) {
  await _shelljs.default.exec(`appium driver uninstall ${driverName}`);
  await _shelljs.default.exec(`appium driver install ${driverName}@${enteredVersion}`);
}
async function retrySelectedDriverInstall(driverName) {
  newLine();
  ui.log.write(_chalk.default.red(`Please try again with a valid version number...`));
  await installDrivers(driverName);
}
async function runAppiumDoctor() {
  const {
    platform
  } = await _inquirer.default.prompt([{
    type: 'list',
    name: 'platform',
    choices: ['android', 'ios', 'dev']
  }]);
  const doctorPath = _path.default.join(__dirname + '/../node_modules/.bin/appium-doctor');
  _shelljs.default.exec(`${doctorPath} --${platform}`);
}
async function installPlugin() {
  const requiredPlugins = await _inquirer.default.prompt([{
    type: 'checkbox',
    message: 'Select Plugins to install',
    name: 'plugins',
    choices: _plugin.plugins,
    validate(answer) {
      if (answer.length < 1) {
        return 'You must choose at least one plugin.';
      }
      return true;
    }
  }]);
  const {
    source
  } = await _inquirer.default.prompt([{
    type: 'list',
    message: 'Source ',
    name: 'source',
    choices: ['npm', 'github', 'git', 'local']
  }]);
  let pluginPath;
  if (source !== 'npm') {
    const path = await _inquirer.default.prompt([{
      name: 'pluginPath',
      message: 'Source of plugin'
    }]);
    pluginPath = path.pluginPath;
  }
  const installedPlugins = _shelljs.default.exec('appium plugin list --installed --json', {
    silent: true
  });
  let pluginNamesInstalled = Object.entries(JSON.parse(installedPlugins.stdout));
  let pluginInformation = [];
  pluginNamesInstalled.map(([key, val]) => {
    // @ts-ignore
    pluginInformation.push({
      pluginName: key,
      plugin: val.pkgName,
      installed: val.installed
    });
  });
  await Promise.all(requiredPlugins.plugins.map(async pluginName => {
    newLine();
    const pluginExists = pluginInformation.find(name => name.plugin === pluginName);
    if (pluginExists !== undefined) {
      ui.log.write(_chalk.default.yellow(`ℹ️  Plugin ${pluginName} already installed`));
      newLine();
      ui.log.write(_chalk.default.yellow(`ℹ️  Checking if any update available for plugin ${pluginName}`));
      newLine();
      _shelljs.default.exec(`appium plugin update ${pluginExists.pluginName}`);
      return;
    } else {
      if (!pluginPath) {
        _shelljs.default.exec(`appium plugin install --source ${source} ${pluginName}`);
      } else {
        _shelljs.default.exec(`appium plugin install --source ${source} --package ${pluginPath} plugin`);
      }
    }
  }));
}
function newLine() {
  ui.log.write('\n');
}
